pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -1757,67 +1767,26 @@ private static class PartitionIdFetcherImpl implements PartitionIdFetcher {
     public int fetch(String segmentName, boolean isConsuming) {
-      Integer partitionId;
-      if (_isStrictRealtimeSegmentAssignment) {
-        // This is how partitionId is calculated for StrictRealtimeSegmentAssignment. Here partitionId is mandatory
-        partitionId =
-            SegmentUtils.getRealtimeSegmentPartitionId(segmentName, _tableNameWithType, _helixManager,
-                _partitionColumn);
-        Preconditions.checkState(partitionId != null, "Failed to find partition id for segment: %s of table: %s",
-            segmentName, _tableNameWithType);
-      } else {
-        TableType tableType = TableNameBuilder.getTableTypeFromTableName(_tableNameWithType);
-        if (tableType == TableType.OFFLINE) {
-          InstancePartitions instancePartitions = _instancePartitionsMap.get(InstancePartitionsType.OFFLINE);
-          assert instancePartitions != null;
-          if (_partitionColumn == null || instancePartitions.getNumPartitions() == 1) {
-            // Fallback to partitionId 0, in this case batching will not be possible so we will fall back to a full
-            // rebalance without batching
-            partitionId = 0;
-          } else {
-            // This is how partitionId is calculated for OFFLINE tables
-            partitionId = SegmentAssignmentUtils.getOfflineSegmentPartitionId(segmentName, _tableNameWithType,
-                _helixManager, _partitionColumn);
-          }
-        } else {
-          if (isConsuming || !_instancePartitionsMap.containsKey(InstancePartitionsType.COMPLETED)) {
-            // This is how partitionId is calculated for CONSUMING segments and ONLINE segments without COMPLETED
-            // instance partitions in RealtimeSegmentAssignment
-            partitionId = SegmentAssignmentUtils.getRealtimeSegmentPartitionId(segmentName, _tableNameWithType,
-                _helixManager, _partitionColumn);
-          } else {
-            // This is how partitionId is calculated for ONLINE segments when COMPLETED instance partitions exist
-            // in RealtimeSegmentAssignment
-            InstancePartitions instancePartitions = _instancePartitionsMap.get(InstancePartitionsType.COMPLETED);
-            assert instancePartitions != null;
-            if (_partitionColumn == null || instancePartitions.getNumPartitions() == 1) {
-              // Fallback to partitionId 0, in this case batching will not be possible so we will fall back to a full
-              // rebalance without batching
-              partitionId = 0;
-            } else {
-              // This is how partitionId is calculated for REALTIME tables if a partition column exists and if the
-              // COMPLETED instance partitions has more than 1 partition
-              partitionId = SegmentAssignmentUtils.getRealtimeSegmentPartitionId(segmentName, _tableNameWithType,
-                  _helixManager, _partitionColumn);
-            }
-          }
-        }
+      Integer partitionId =
```

Comment:
you should handle the `numPartitions = 1` case here, right? to keep it consistent with how the partitionId is assigned for replica group assignment? Which also means you need to use the `isConsuming` flag as the original code uses it?

Reply:
From abstraction, IMO the fetcher should just read the partition id from segment regardless of `InstancePartitions.numPartitions`.
In this step, we simply calculate the partition id associated with the segment. In the following step, we will assign segments based on `InstancePartitions`.
Maybe I missed something?

Reply:
discussed offline that it is okay to keep the `InstancePartitions.numPartitions` check decoupled from rebalance perspective. If someone is using strict replica group routing with just 1 partition or with partition column as null, that is not really the correct way to use this (and in the future we may address this via validations etc)

I will do some rebalance testing with the changes and get back with findings